### PR TITLE
Harden security - stage two

### PR DESCRIPTION
**Pull request by mindplay-dk**
Following up on [this PR](https://github.com/kodus/session/pull/6), this PR continues to harden security in other ways.

After careful contemplation and discussion, @boan2010 and I have decided to follow this recommendation:

https://stackoverflow.com/a/12234563/283851

In other words, we're *not* pursuing any means of further hardening the cookie-value itself (using hashes or timestamps etc.) which is an ultimately futile effort. (read the above post for details.)

We're also *not* performing any periodic renewal of the session ID, and this is discouraged in the documentation - the conclusion was that the *only* time you can safely renew a session ID is immediately after verifying user credentials, e.g. at the time of login. Renewing the session at any other time just enables a hijacker to "fork" your session. This is likely fine for "medium security" applications - whereas "higher security" applications should use a shorter TTL, e.g. short enough to ensure most session will expire overnight and force the user to log in daily.

We do want to offer the ability to renew the session, however - so, in a login-controller, we want the ability to destroy the previous Session ID and issue a new one, while retaining session state. This feature has been added via the `Session::renew()` method.

We've also separate the concept of TTL from the expiration of the cookie itself:

TTL now means "time to live *of the session data in storage*", where previously this also meant "time to live of the cookie".

Expiration of the cookie itself is now much longer - one year by default. This does *not* mean the session data lives for that long, but it does means that you can keep renewing it for that long. (We referenced google.com, which persists the session ID similarly in a cookie with a two-year expiration.)

The cookie itself (as previously) does *not* get a renewed expiration per request - if the user never logs in/out (on a site with a TTL that permits the user to keep updating the session) they *will* experience a "random" session-termination after a year, which seems acceptable; and complies with the OWASP recommendation that a session cookie must have an absolute expiration date.

This PR also adds support for the `Domain` cookie-attribute. By default, we do not set this attribute, which means our session-cookie is *not* accessible from sub-domains by default. You can read more about the `Domain` attribute here:

https://stackoverflow.com/questions/1062963/how-do-browser-cookie-domains-work/30676300#30676300

I've also added some missing documentation and tests.

---

If this PR is approved, `kodus/user-session` (et al) will need updates to take advantage of session renewal.

This PR will not be merged before @rhl-jfm returns from vacation and has a chance to review it.

In order to support higher security (shorter TTL) for back-end/admin pages, we may still need to think more about supporting multiple sessions via separate cookies. This issue will need further discussion and likely a separate PR.

One additional outstanding issue for discussion is the OWASP recommendation that Session IDs should not be stored plainly in the database. The `kodus/user-session` module does that. (This issue doesn't directly relate to `kodus/session`, but an additional enhancement might make sense to support the use-case.)

We also need to decide how to version this. There are no breaking changes to the API, but some of the changes to the behavior could be considered "breaking"?

### CODE REVIEW

**Review by thomasnordahl-dk — APPROVED** [review]
Seems like you hit the nail on the head :)

File: src/SessionService.php [hunk 1]
```diff
@@ -23,6 +23,11 @@ class SessionService
      */
     const TWO_WEEKS = 1209600;
 
+    /**
+     * @var int one year in seconds
+     */
+    const ONE_YEAR = 31536000;
```
Comment (boan-jfm) [inline src/SessionService.php]: You can add visibility modifiers in PHP 7.1 to constants
http://php.net/manual/en/language.oop5.constants.php#language.oop5.basic.class.this

**Review by boan-jfm — APPROVED** [review]

File: src/SessionService.php (hunk 1 above)
Comment (mindplay-dk) [reply src/SessionService.php]: That's a breaking change - at the moment, the API is backwards-compatible.

**Comment by mindplay-dk** [comment]
Note that I could add more strict type-hints around the codebase, as well as adjusting visibility of some public members - I haven't done so because I'm maintaining backwards compatibility.

We still need to decide whether the next release can be considered *functionally* backwards compatible, even if the API is compatible - if we decide it's a major release, I could definitely take another pass at type-hints and visibility modifiers.

**Comment by boan-jfm** [comment]
It is set to >= PHP 7.1 in the composer.json:

https://github.com/kodus/session/blob/master/composer.json#L19

**Comment by mindplay-dk** [comment]
If we `clear()` the session, it gets internally renewed - but we're currently not invalidating the old Session ID if the session is empty when we commit.

This needs to be addressed.

**Comment by mindplay-dk** [comment]
> One additional outstanding issue for discussion is the OWASP recommendation that Session IDs should not be stored plainly in the database

The `getSessionID()` method should return a *hashed* Session ID, and this should be documented.

It must be formatted as a GUID v4 per the spec.

An internal `getInternalSessionID()` method should return the internal Session ID, which is the value that gets written to the cookie.

**Comment by mindplay-dk** [comment]
> There are no breaking changes to the API, but some of the changes to the behavior could be considered "breaking"?

The cookie value isn't backwards compatible, so we'll consider this a major release.

We'll go ahead and add PHP 7 type-hints as needed.

**Comment by mindplay-dk** [comment]
> We'll go ahead and add PHP 7 type-hints as needed.

Everything was already type-hinted for PHP 7.0 in the 2.0 release - there are no new type-hinting features in PHP 7.1 applicable to the codebase, so we'll keep this as-is.

I'm not privatizing the consts either, as there's no particular reason to do so - they're already part of the public API, they're required by the tests, and they may be useful in other contexts.

**Comment by boan-jfm** [comment]
> I'm not privatizing the consts either, as there's no particular reason to do so - they're already part of the public API, they're required by the tests, and they may be useful in other contexts.

We are just explicit about the visibility of our constants e.g. `public const` instead of just `const`. You could also be implicit about [class methods](https://www.tehplayground.com/cEnc6wqPaU4kXHpi), they also defaults to public :-)

But okay...

**Comment by mindplay-dk** [comment]
@boan2010 sorry, I thought you were suggesting we privatize the consts - yes, I'll add `public` :-)

**Pull request merged**